theseus/optimizer/nonlinear/nonlinear_optimizer.py

```diff
@@ -191,43 +202,130 @@ def _optimize_impl(
+        self,
+        track_best_solution: bool = False,
+        verbose: bool = False,
+        backward_mode: BackwardMode = BackwardMode.FULL,
+        **kwargs,
+    ) -> OptimizerInfo:
+        with torch.no_grad():
+            info = self._init_info(track_best_solution, verbose)
+
+        if verbose:
+            print(
+                f"Nonlinear optimizer. Iteration: 0. "
+                f"Error: {info.last_err.mean().item()}"
+            )
+
+        if backward_mode == BackwardMode.FULL:
+            return self._optimize_loop(
+                start_iter=0,
+                num_iter=self.params.max_iterations,
+                info=info,
+                track_best_solution=track_best_solution,
+                verbose=verbose,
+                force_update=False,
+                **kwargs,
+            )
+        elif backward_mode in [BackwardMode.IMPLICIT, BackwardMode.TRUNCATED]:
+            if backward_mode == BackwardMode.IMPLICIT:
+                backward_num_iterations = 1
+            else:
+                if "backward_num_iterations" not in kwargs:
+                    raise ValueError()
+                backward_num_iterations = kwargs["backward_num_iterations"]
+
+            num_no_grad_iter = self.params.max_iterations - backward_num_iterations
+            with torch.no_grad():
+                self._optimize_loop(
+                    start_iter=0,
+                    num_iter=num_no_grad_iter,
+                    info=info,
+                    track_best_solution=track_best_solution,
+                    verbose=verbose,
+                    force_update=False,
+                    **kwargs,
+                )
+
+            grad_loop_info = self._init_info(track_best_solution, verbose)
+            self._optimize_loop(
+                start_iter=0,
+                num_iter=backward_num_iterations,
+                info=grad_loop_info,
+                track_best_solution=False,
+                verbose=verbose,
+                force_update=True,
+                **kwargs,
+            )
+
+            if any(grad_loop_info.error_increase_indices):
+                raise RuntimeError("Grad loop error increased")
+
+            # Merge the converged status into the info from the detached loop,
+            # and for now, don't update the best err tracking or best solution.
+            M = info.status == NonlinearOptimizerStatus.MAX_ITERATIONS
+            assert np.all(
+                (grad_loop_info.status[M] == NonlinearOptimizerStatus.MAX_ITERATIONS)
+                | (grad_loop_info.status[M] == NonlinearOptimizerStatus.CONVERGED)
+            )
+            info.status[M] = grad_loop_info.status[M]
+            info.converged_indices[M] = grad_loop_info.converged_indices[M]
+            info.converged_iter[M] = (
+                info.converged_iter[M] + grad_loop_info.converged_iter[M]
+            )
+
+            return info
+        else:
+            raise ValueError()
```

Comment:
Add error message along the lines of unknown backward mode.

Reply:
Added